File: CSharp/ImageProcessing/ImageProcessing/Binarization.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Imaging;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Windows;

/// <summary>
/// 2値化のロジック
/// </summary>
class Binarization
{
    /// <summary>
    /// コンストラクタ
    /// </summary>
    public Binarization()
    {
    }

    /// <summary>
    /// デスクトラクタ
    /// </summary>
    ~Binarization()
    {
    }

    /// <summary>
    /// 2値化の実行
    /// </summary>
    /// <param name="bitmap">ビットマップ</param>
    /// <param name="token">キャンセルトークン</param>
    /// <param name="thresh">閾値</param>
    /// <returns>実行結果 成功/失敗</returns>
    public bool ImageProcessing(ref Bitmap bitmap, CancellationToken token, byte thresh)
    {
        bool result = true;

        int widthSize = bitmap.Width;
        int heightSize = bitmap.Height;

        var bitmapData = bitmap.LockBits(new Rectangle(0, 0, widthSize, heightSize), ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb);

        int indexWidth;
        int indexHeight;

        unsafe
        {
            for (indexHeight = 0; indexHeight < heightSize; indexHeight++)
            {
                if (token.IsCancellationRequested)
                {
                    result = false;
                    break;
                }

                for (indexWidth = 0; indexWidth < widthSize; indexWidth++)
                {
                    if (token.IsCancellationRequested)
                    {
                        result = false;
                        break;
                    }

                    byte* pixel = (byte*)bitmapData.Scan0 + indexHeight * bitmapData.Stride + indexWidth * 4;
                    byte grayScale = (byte)((pixel[(int)ComInfo.Pixel.B] + pixel[(int)ComInfo.Pixel.G] + pixel[(int)ComInfo.Pixel.R]) / 3);

                    byte binarization = grayScale >= thresh ? (byte)255 : (byte)0;
                    pixel[(int)ComInfo.Pixel.B] = binarization;
                    pixel[(int)ComInfo.Pixel.G] = binarization;
                    pixel[(int)ComInfo.Pixel.R] = binarization;
                }
            }
            bitmap.UnlockBits(bitmapData);
        }

        return result;
    }
}

File: Vb/ImageProcessing/ImageProcessing/Binarization.vb
﻿Imports System.Threading
Imports System.Runtime.InteropServices.Marshal
Imports System.Drawing.Imaging

''' <summary>
''' 2値化のロジック
''' </summary>
Public Class Binarization
    ''' <summary>
    ''' コンストラクタ
    ''' </summary>
    Public Sub New()
    End Sub

    ''' <summary>
    ''' デスクトラクタ
    ''' </summary>
    Protected Overrides Sub Finalize()
        MyBase.Finalize()
    End Sub

    ''' <summary>
    ''' 2値化の実行
    ''' </summary>
    ''' <param name="bitmap">ビットマップ</param>
    ''' <param name="token">キャンセルトークン</param>
    ''' <param name="thresh">閾値</param>
    ''' <returns>実行結果 成功/失敗</returns>
    Public Function ImageProcessing(ByRef bitmap As Bitmap, token As CancellationToken, thresh As Byte) As Boolean
        Dim result = True

        Dim widthSize = bitmap.Width
        Dim heightSize = bitmap.Height

        Dim bitmapData = bitmap.LockBits(New Rectangle(0, 0, widthSize, heightSize), ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb)

        Dim indexWidth As Integer
        Dim indexHeight As Integer

        For indexHeight = 0 To heightSize - 1 Step 1
            If (token.IsCancellationRequested = True) Then
                result = False
                Exit For
            End If

            For indexWidth = 0 To widthSize - 1 Step 1
                If (token.IsCancellationRequested = True) Then
                    result = False
                    Exit For
                End If

                Dim position = indexHeight * bitmapData.Stride + indexWidth * 4
                Dim pixelB = ReadByte(bitmapData.Scan0, position + ComInfo.Pixel.B)
                Dim pixelG = ReadByte(bitmapData.Scan0, position + ComInfo.Pixel.G)
                Dim pixelR = ReadByte(bitmapData.Scan0, position + ComInfo.Pixel.R)

                Dim grayScale = (pixelB + pixelG + pixelR) / 3

                Dim binarization = If(grayScale >= thresh, CByte(255), CByte(0))
                WriteByte(bitmapData.Scan0, position + ComInfo.Pixel.B, CByte(binarization))
                WriteByte(bitmapData.Scan0, position + ComInfo.Pixel.G, CByte(binarization))
                WriteByte(bitmapData.Scan0, position + ComInfo.Pixel.R, CByte(binarization))
            Next
        Next
        bitmap.UnlockBits(bitmapData)

        Return result
    End Function
End Class